### PR TITLE
Adds meet link to scheduled interviews and sends emails to Shadow

### PR DESCRIPTION
**Pull request by GrantComm**
This pr includes the meet link when emails are sent to the interview participants. It also sends emails to the shadow of an interview. 

Screenshot 2020-08-06 at 5 25 05 PM

Screenshot 2020-08-06 at 5 25 46 PM

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 1]
```diff
@@ -137,7 +159,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     InterviewPostOrPutRequest postRequest;
     try {
       postRequest = new Gson().fromJson(getJsonString(request), InterviewPostOrPutRequest.class);
-    } catch (Exception JsonSyntaxException) {
+    } catch (JsonSyntaxException jse) {
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:162]: Good catch, I guess I missed this one.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 2]
```diff
@@ -232,19 +263,26 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
   public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String shadowEmail = userService.getCurrentUser().getEmail();
     String shadowId = getUserId();
+
     InterviewPostOrPutRequest putRequest;
     try {
       putRequest = new Gson().fromJson(getJsonString(request), InterviewPostOrPutRequest.class);
-    } catch (JsonSyntaxException jse) {
+    } catch (Exception JsonSyntaxException) {
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Why is this reverted? Please change back to JsonSyntaxException jse

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 3]
```diff
@@ -253,32 +291,46 @@ public void doPut(HttpServletRequest request, HttpServletResponse response) thro
...
+
     List<ScheduledInterview> possibleInterviews =
         ShadowLoadInterviewsServlet.getPossibleInterviews(
             scheduledInterviewDao, selectedPosition, interviewRange, personDao, shadowId);
+
     Set<ScheduledInterview> notValidInterviews = new HashSet<ScheduledInterview>();
     // We want to remove all interviews where the company or job does not match that
     // specified in the request.
-    String interviewerCompany = putRequest.getCompany();
-    String interviewerJob = putRequest.getJob();
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Please put these back where they were, Alex wants them as close to where they're used as possible.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 4]
```diff
@@ -430,8 +484,20 @@ private static String formatPositionString(String str) {
...
+      String meetLink, ScheduledInterview previousInterviewObject) {
+    return ScheduledInterview.create(
+        previousInterviewObject.id(),
+        previousInterviewObject.when(),
+        previousInterviewObject.interviewerId(),
+        previousInterviewObject.intervieweeId(),
+        meetLink,
+        previousInterviewObject.position(),
+        previousInterviewObject.shadowId());
+  }
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: This shouldn't be necessary anymore with the withMeetLink function.

File: src/main/webapp/scheduled-interviews.jsp [hunk 5]
```diff
@@ -23,16 +23,25 @@
             <li class="list-group-item">Interviewee: ${scheduledInterview.getInterviewee()}</li>
             <li class="list-group-item">Interviewer: ${scheduledInterview.getInterviewer()}</li>
             <li class="list-group-item">Shadow: ${scheduledInterview.getShadow()}</li>
+            <li class="list-group-item">
+              <a href=${scheduledInterview.getMeetLink()} target="_blank">
+                MeetLink
```
Comment (meagancheese) [inline src/main/webapp/scheduled-interviews.jsp]: Nit: add space between Meet and Link.

File: src/test/java/com/google/sps/servlets/ScheduledInterviewServletTest.java [hunk 6]
```diff
@@ -374,7 +378,7 @@ public void onlyReturnsInterviewersWithMatchingCompanyAndJob() throws IOExceptio
                 Instant.parse("2020-07-20T12:45:00Z"), Instant.parse("2020-07-20T13:45:00Z")),
             googleSWE1.id(),
             emailToId("user@company.org"),
-            "meet_link",
+            String.valueOf(actual.get(0).id()),
```
Comment (meagancheese) [inline src/test/java/com/google/sps/servlets/ScheduledInterviewServletTest.java:381]: Why is this like this?

**Review by meagancheese — CHANGES_REQUESTED** [review]
Some small things and a question.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 2 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done, happened while resolving conflicts

File: src/test/java/com/google/sps/servlets/ScheduledInterviewServletTest.java (hunk 6 above)
Comment (GrantComm) [reply src/test/java/com/google/sps/servlets/ScheduledInterviewServletTest.java:381]: The mock meetlink is the id of the scheduled interview as a string

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 4 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/test/java/com/google/sps/servlets/ScheduledInterviewServletTest.java (hunk 6 above)
Comment (meagancheese) [reply src/test/java/com/google/sps/servlets/ScheduledInterviewServletTest.java:381]: Ok, that's fine then.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 3 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

**Review by meagancheese — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 7]
```diff
@@ -166,8 +188,10 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         getPossibleInterviewerIds(allAvailableInterviewers, interviewerCompany, interviewerJob);
     int randomNumber = (int) (Math.random() * possibleInterviewers.size());
     String interviewerId = possibleInterviewers.get(randomNumber);
-    // TODO: replace with real MeetLink from Calendar Access
-    String meetLink = "meet_link";
+
+    // Meet link is empty because we need a scheduledInterview before the link can be created
+    String meetLink = "";
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Why don't you just put this into the creation of the SchInt? Updating it after creation doesn't fit what the other fields are doing

**Review by apluscs — CHANGES_REQUESTED** [review]
1 Nit/question

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 7 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 8]
```diff
@@ -232,6 +260,7 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
   public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String shadowEmail = userService.getCurrentUser().getEmail();
     String shadowId = getUserId();
+
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: What's with all these blank lines?

File: src/main/java/com/google/sps/data/SendgridEmailSender.java [hunk 9]
```diff
@@ -24,6 +24,12 @@
 import com.sendgrid.helpers.mail.objects.Content;
 import com.sendgrid.helpers.mail.objects.Email;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
```
Comment (acmcarther) [inline src/main/java/com/google/sps/data/SendgridEmailSender.java]: Why add these imports?

Comment (GrantComm) [reply src/main/java/com/google/sps/data/SendgridEmailSender.java]: Added by accident when merging

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 8 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: used them to see code clearer, just removed

**Review by acmcarther — APPROVED** [review]

**Pull request merged**